### ISSUE TITLE
Entity version metadata graph provenance is not well represented

### ISSUE THREAD

**Issue by diekhans** [HumanCellAtlas/metadata-schema#841]
## Problem

In constructing a POC of project level metadata JSON file, I came across provenance issues demonstrating that constructing the version metadata graph was going to be extremely difficult.  
This relates to the versions of the entities, not the schema versions.

The POC project is being used as a use case for consumers of HCA, especially red-box developers, but also biologist-with a computer.  The UX group believes this represents a real-world HCA use case, not use a one-off internal problem.

Data from both links.json and the bundle manifest must be merged with the entities to understand the provenance.

The issues revolve around lost of version and difficult in traversing that graph when considering that
entities can change.  

- Metadata entities don't contain their version number or full-qualified id.  One can only determine the version number of a metadata entity by looking at the bundle-manifest.  With a goal of ease of understanding, the version would be contained in the entity, not an external file that is not part of the metadata data model.

- References between entities are contained in links.json, which has no provenance information at all.  All of the links are by simple UUID, not FQID, making the provenance history difficult to track and understanding it directly tied to the DSS.

Any attempt to construct the versioned graph would be very difficult, essentially obtaining previous versions of each bundle and reconstructing the provenance.  This highly confusing, and will be both difficult to code and likely result in provenance just being ignored by any application developer.  Anyone wanting to understand changes to the data or metadata will be impacted.   This is important to understand, as the way analysis is done can change based on metadata, resulting in different downstream results.

## Proposal

I believe we should modify the entity model to represent the provenance independent of the DSS and in a way that  is obvious and easy to use.  I think this is important because it will be hard to retrofit.

Here is a straw-man proposal on a straight forward representation based on the POC work.

- system/provenance adds the following fields:
  - document_fqid - the fully qualified id of the current version of document, that is uuid.version. Storing the version as an independent field is also possible and we might want to do redundantly.  FQIDs are easier to use in lists of references.
  - document_previous_fqid - the fully qualified id of the previous version of document, uuid.version.
  - input_fqids - the back-links in the DAG.  This allows the DAG to be constructed bottom-up, including full version history.  Forward DAG links are not possible, one can't go back and modify 
the previous entity in the graph without changing its version.
- process/process_core as the following field:
  - protocol_fqids - list of protocol FQIDS used by the process

At this point, links.json is redundant, however, it probably should be maintained for compatibility and may be easier for some to use.  Following changes are suggested to links.json, although I have not tested them in the POC):
- add fqid references for all entities to in parallel to the current unqualified ids.
- add standard provenance information

The above does not represent how to link project objects into the graph.

**Comment by diekhans** [comment]
Related to #511

**Comment by simonjupp** [comment]
Ironically the links are represented on entities when you access them from the ingest API. e.g. http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290 contains a `_links` section. 

```
"_links": {
"self": {
"href": "http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290"
},
"biomaterial": {
"href": "http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290{?projection}",
"templated": true
},
"processing": {
"href": "http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290/processingEvent"
},
"draft": {
"href": "http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290/draftEvent"
},
"projects": {
"href": "http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290/projects",
"title": "Access or create projects. Creation can only be done inside a submission envelope"
},
"inputToProcesses": {
"href": "http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290/inputToProcesses"
},
"derivedByProcesses": {
"href": "http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290/derivedByProcesses"
},
"submissionEnvelopes": {
"href": "http://api.ingest.data.humancellatlas.org/biomaterials/5ab918367b6ab0000753d290/submissionEnvelopes",
"title": "Access or create new submission envelopes"
}
}
```

The links are removed when we export to the DSS as they point to absolute URLs in ingest. But if we have access to the fqids at export then maybe we could rewrite the links to point to their absolute URLs in the DSS. 

Using canonical URIs over fqids would be more in line with best practices for publishing linked data and removes ambiguity for consumers. Getting file content becomes a `GET` requests rather than having to scan potentially thousands of metadata files that have been downloaded to disk.

**Comment by lauraclarke** [comment]
+1 to @simonjupp suggestion @kozbo @xbrianh do you see this working?

**Comment by xbrianh** [comment]
@lauraclarke Yes, passing in references to DSS objects should work fine.

**Comment by lauraclarke** [comment]
Great, sounds like @simonjupp, @justincc and @diekhans might want to work up an example and make sure it looks sane to everyone and then we have a way forward to resolve this issue.

**Comment by diekhans** [comment]
I am not sure how using URIs will work when the metadata can be in multiple locations.  AWS, GCP, my hard disk.  Well, they could be treated as the ID of the object, not the location.

**Comment by diekhans** [comment]
Here is an example of the changes I made to the provenance records in the POC.

<pre>
       "provenance": {
            "document_fqid": "3efd3ea4-a535-4560-960a-773dc7f85029.2018-12-05T053830.196000Z",
            "document_id": "3efd3ea4-a535-4560-960a-773dc7f85029",
            "document_version": "2018-12-05T053830.196000Z",
            "inputs": [
                "1d2effc1-5f4b-41d6-aa28-cc17a956650d.2018-12-04T191049.794000Z"
            ],
            "submission_date": "2018-12-04T18:53:44.858Z",
            "update_date": "2018-12-05T05:38:30.196Z"
        },
 </pre>

**Comment by diekhans** [comment]
@simonjupp could you make up a concrete example of how the links would look in the provenance sections?!?!

**Comment by simonjupp** [comment]
In HAL it would look something like this. 
```
 "_embedded" {
       "provenance": {
            "document_fqid": "3efd3ea4-a535-4560-960a-773dc7f85029.2018-12-05T053830.196000Z",
            "document_id": "3efd3ea4-a535-4560-960a-773dc7f85029",
            "document_version": "2018-12-05T053830.196000Z",
            "submission_date": "2018-12-04T18:53:44.858Z",
            "update_date": "2018-12-05T05:38:30.196Z"
        }
},
 "_links": {
   "self": {
       "href": "https://dss.prod.data.humancellatlas.org/files/3efd3ea4-a535-4560-960a-773dc7f85029?version=2018-12-05T053830.196000Z",
 },
   "inputs":  [
   {"href": "https://dss.prod.data.humancellatlas.org/files/1d2effc1-5f4b-41d6-aa28-cc17a956650d?version=2018-12-04T191049.794000Z"}
  ]
}
```

**Comment by diekhans** [comment]
thanks @simonjupp 

Is the issue really links vs ids?  I already have the JSON, I want to know the version without figuring out which bundle to look at.  Parsing ids from URLs is fragile  Calling a web API with ids seems routine.  I could see having both URLs and IDs being helpful, but a different discussion.

URLs seems to create interesting issues with location unrelated to identification.  Like what if the entity isn't in production yet?  Do you rewrite the URLs?  or does the entity get edited when moved,
hence creating a new version?

Right now my concern is having a versioned identification model that is contained within the metadata and doesn't require understanding which bundle you happened to get that copy of the metadata entity from.

links.json actually makes this harder, not easier.  I am not suggesting getting rid of it, just not making it necessary.

**Comment by lauraclarke** [comment]
@diekhans just to confirm, while what Simon has proposed does give us links inside documents, another requirement you have spotted, is the ability to determine what schema version a current document is and this solution does not solve that problem

**Comment by daniwelter** [comment]
> Related to #511

@diekhans This issue is actually related to metadata schema versions rather than documents. I've updated the issue accordingly. 

Regarding the proposal above, the one thing that would concern me is the proposed modification by ingest of a non-system schema entity, namely process_core. The provenance.json schema is designed to be entirely added and modified by the ingest system, so that part is fine, but no other part of the metadata should be touched by ingest once the JSON doc has been created. IIRC, the linked graph isn't built until until after the json docs have been created, so the system would have to go back and modify validated JSON docs after this stage. I would caution against this.

**Comment by simonjupp** [comment]
> thanks @simonjupp
> 
> Is the issue really links vs ids? I already have the JSON, I want to know the version without figuring out which bundle to look at. Parsing ids from URLs is fragile Calling a web API with ids seems routine. I could see having both URLs and IDs being helpful, but a different discussion.

Yeah, so in this scenario where you already have all the JSON in one file, the URIs are less relevant. I'm thinking more when you access a `file` via the DSS API. In that scenario I think HAL links are more useful for clients that are crawling the API for metadata.   Ids like `1d2effc1-5f4b-41d6-aa28-cc17a956650d.2018-12-04T191049.794000Z`  need out of bounds information to work out how to access the resource. 

I think `protocol_fqids` could go in `process/provenance` rather than `process_core`.

I'd also suggest updating links.json to include the fqids if we're keeping it.

**Comment by diekhans** [comment]
Another related issue is that the project is not linked into the metadata graph in this proposal.  The relationship of project with other entities is not clear and only implicit via bundle.

**Comment by zperova** [comment]
@lauraclarke to bring this up on a meeting to discuss further
